src/intfuncs.c

```diff
@@ -617,6 +617,211 @@ Obj FuncINT_STRING ( Obj self, Obj string )
+                     0);
+    UInt x = INT_INTOBJ(data);
+    UInt y = INT_INTOBJ(val);
+    return INTOBJ_INT((x & ~mask) | (y << offset));
+}
+
+static Obj DoBooleanFieldGetter(Obj self, Obj data)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+    if (!IS_INTOBJ(data))
+        ErrorMayQuit("Boolean Field getter: argument must be small integer", 0, 0);
+    UInt x = INT_INTOBJ(data);
+    return (x & mask) ? True : False;
+}
+
+static Obj DoBooleanFieldSetter(Obj self, Obj data, Obj val)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+    if (!IS_INTOBJ(data))
+        ErrorMayQuit("Boolean Field Setter: data must be small integer", 0,
+                     0);    
+    UInt x = INT_INTOBJ(data);
+    if (val == True)
+      x |= mask;
+    else if (val == False)
+      x &= ~mask;
+    else
+      ErrorMayQuit("Boolean Field Setter: value must be true or false", 0, 0);
+    return INTOBJ_INT(x);
+}
+
+
+static Obj FuncBUILD_BITFIELDS(Obj self, Obj args)
+{
+  GAP_ASSERT(IS_PLIST(args));
+  GAP_ASSERT(LEN_PLIST(args) >= 1  && ELM_PLIST(args,1));
+  Obj widths = ELM_PLIST(args,1);
+  if (!IS_LIST(widths))
+    ErrorMayQuit("Fields builder: first argument must be list of widths", 0,
+                     0);
+  UInt nfields = LEN_LIST(widths);
+  if (LEN_PLIST(args) != nfields + 1)
+    ErrorMayQuit("Fields builder: number of values must match number of widths", 0, 0);
+  UInt x = 0;
+  UInt i;
+  for (i = nfields; i > 0; i--) {
+    GAP_ASSERT(ISB_LIST(widths, i));
+    Obj y = ELM_LIST(widths,i);
+    GAP_ASSERT(IS_INTOBJ(y));
+    x <<= INT_INTOBJ(y);
+    GAP_ASSERT(ELM_PLIST(args, i+1));
+    Obj z = ELM_PLIST(args, i+1);
+    if (!IS_INTOBJ(z))
+      ErrorMayQuit("Fields builder: values must be small integers", 0, 0);
+    GAP_ASSERT(INT_INTOBJ(z) < (1 << INT_INTOBJ(y)));
+    x |= INT_INTOBJ(z);
+  }
+  return INTOBJ_INT(x);
+}
+
+
+Obj FuncMAKE_BITFIELDS(Obj self, Obj widths)
+{
+    if (!IS_LIST(widths))
+        ErrorMayQuit("MAKE_BITFIELDS: widths must be a list", 0, 0);
+    UInt nfields = LEN_LIST(widths);
+    UInt starts[nfields + 1];
+    starts[0] = 0;
+    for (UInt i = 1; i <= nfields; i++) {
+        Obj o = ELM_LIST(widths, i);
+        if (!IS_INTOBJ(o))
+            ErrorMayQuit("MAKE_BITFIELDS: widths must be small integers", 0,
+                         0);
+        UInt width = INT_INTOBJ(o);
+        starts[i] = starts[i - 1] + width;
+    }
+    if (starts[nfields] > 8 * sizeof(UInt))
+        ErrorMayQuit("MAKE_BITFIELDS: total widths too large", 0, 0);
+
+    Obj  setters = NEW_PLIST(T_PLIST_DENSE + IMMUTABLE, nfields);
+    Obj  getters = NEW_PLIST(T_PLIST_DENSE + IMMUTABLE, nfields);
+    Obj  bsetters = NEW_PLIST(T_PLIST + IMMUTABLE, nfields);
+    UInt bslen = 0;
+    Obj  bgetters = NEW_PLIST(T_PLIST + IMMUTABLE, nfields);
+    for (UInt i = 1; i <= nfields; i++) {
+        UInt mask = (1L << starts[i]) - (1L << starts[i - 1]);
+        Obj  s = NewFunctionCT(T_FUNCTION, SIZE_FUNC, "<field setter>", 2,
```

Comment:
You could also just `NewFunctionC` here and elsewhere (makes it slightly easier to read the code, IMO, but that's perhaps taste)